src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Grammar/GrammarGenerator.cs

```diff
@@ -0,0 +1,325 @@
+        private readonly Dictionary<string, List<Production>> _nameToProductions;
+
+        public GrammarGenerator(Tree tree)
+        {
+            // Syntax refers to a special pseudo-element 'Modifier'.  Just synthesize that since
+            // it's useful in the g4 grammar.
+            var modifierKeywords = from mod in GetKinds<DeclarationModifiers>()
+                                   let modKeyword = mod + "Keyword"
+                                   from kind in SyntaxKinds
+                                   where modKeyword == kind.ToString()
+                                   select new Kind { Name = modKeyword };
+
+            tree.Types.Add(new Node
+            {
+                Name = "Modifier",
+                Children = { new Field { Type = SyntaxToken, Kinds = modifierKeywords.ToList() } }
+            });
+
+            _nodes = tree.Types.Where(t => t.Name != "CSharpSyntaxNode").ToImmutableArray();
+            _nameToProductions = _nodes.ToDictionary(n => n.Name, _ => new List<Production>());
+        }
+
+        public string Run()
+        {
+            foreach (var node in _nodes)
+            {
+                // If this node has a base-type, then have the base-type point to this node as a
+                // valid production for itself.
+                if (node.Base is string nodeBase && _nameToProductions.TryGetValue(nodeBase, out var baseProductions))
+                    baseProductions.Add(RuleReference(node.Name));
+
+                if (node is Node)
+                {
+                    if (node.Children.Count == 0)
+                        continue;
+
+                    // Convert a rule of `a: (x | y | z)` into:
+                    // a: x
+                    //  | y
+                    //  | z;
+                    if (node.Children.Count == 1 && node.Children[0] is Field field && field.IsToken)
+                    {
+                        ProcessProductions(node, field.Kinds.Select(k =>
+                            new List<TreeTypeChild> { new Field { Type = SyntaxToken, Kinds = { k } } }).ToArray());
+                    }
+                    else
+                    {
+                        ProcessProductions(node, node.Children);
+                    }
+                }
+            }
+
+            // The grammar will bottom out with certain lexical productions.  Just emit a few empty
+            // productions in the grammar file indicating what's going on, and making it so that the
+            // g4 file is considered legal (i.e. no rule references names of rules that don't
+            // exist).
+
+            foreach (var kind in s_lexicalTokens)
+                _nameToProductions.Add(kind.ToString(), new List<Production> { new Production("/* see lexical specification */") });
+
+            return GenerateResult();
+        }
+
+        private void ProcessProductions(TreeType node, params List<TreeTypeChild>[] productions)
```

Comment:
Consider making a local function and naming `addChildren`. `productions` is confusing since we're not passing in the `Production` type

Reply:
yeah, changed name here.